x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/utils/get_current_rule_statuses.ts

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { chunk } from 'lodash';
+import { initPromisePool } from '../../../../../utils/promises_pool';
+import { GetCurrentStatusBulkResult, IRuleExecutionLogClient } from '../../../rule_execution_log';
+
+const RULES_PER_CHUNK = 1000;
+
+/**
+ * Get the most recent execution status for each of the given rule IDs.
+ * This method splits work into chunks so not to owerwhelm Elasticsearch
+ * when fetching statuses for a big number of rules.
+ *
+ * @param ruleIds Rule IDs to fetch statuses for
+ * @param execLogClient RuleExecutionLogClient
+ * @param spaceId Current Space ID
+ * @returns A dict with rule IDs as keys and rule statuses as values
+ *
+ * @throws AggregateError if any of the rule status requests fail
+ */
+export async function getCurrentRuleStatuses(
+  ruleIds: string[],
+  execLogClient: IRuleExecutionLogClient,
+  spaceId: string
+): Promise<GetCurrentStatusBulkResult> {
+  const { results, errors } = await initPromisePool({
+    concurrency: 1,
+    items: chunk(ruleIds, RULES_PER_CHUNK),
+    executor: (ruleIdsChunk) =>
+      execLogClient.getCurrentStatusBulk({
+        ruleIds: ruleIdsChunk,
+        spaceId,
+      }),
+  });
+
+  if (errors.length) {
+    throw new AggregateError(errors, 'Error fetching rule statuses');
+  }
```

Comment:
Is it worth returning the partial results and an `errors[]` like we do with other bulk operations so it's not a full on failure if there's an issue with just one of the rule status SO's?

Reply:
++ rule monitoring data is not critical comparing to the rule data itself, and I agree that it would be worth returning partial results from the API.

I'd probably even return any found results as is without returning or re-throwing errors. Errors could be logged via the console logger instead.

Reply:
I don't think that would be a good experience from a user's perspective. We should communicate that the rule monitoring table has missing data due to fetching errors. Otherwise, it would confuse users and appear as rules are not running.

Considering that the fix is for 8.0, I would prefer to preserve the current behavior. We can discuss further improvements to the error handling logic separately.

Reply:
> Otherwise, it would confuse users and appear as rules are not running.

That's true, but if we just throw an exception and don't handle it in the endpoint, the cost for the user would be inability to work with Rule Management and Monitoring tables, so... 🤷 

> We should communicate that the rule monitoring table has missing data due to fetching errors.

I think Garrett's proposal is a middle ground: we'd return rules and still be able to communicate errors in fetching their statuses. 

> Considering that the fix is for 8.0, I would prefer to preserve the current behavior. We can discuss further improvements to the error handling logic separately.

Works for me 👍

Reply:
Added an item for this improvement to "Stability and performance" section of the [epic](https://github.com/elastic/kibana/issues/118324).